core/trie/bitarray.go

```diff
@@ -0,0 +1,458 @@
+		b.words[1] = (b.words[1] >> n) | (b.words[2] << (64 - n))
+		b.words[2] = (b.words[2] >> n) | (b.words[3] << (64 - n))
+		b.words[3] >>= n
+	}
+
+	return b
+}
+
+// Sets b = x ^ y and returns b.
+func (b *BitArray) Xor(x, y *BitArray) *BitArray {
+	b.words[0] = x.words[0] ^ y.words[0]
+	b.words[1] = x.words[1] ^ y.words[1]
+	b.words[2] = x.words[2] ^ y.words[2]
+	b.words[3] = x.words[3] ^ y.words[3]
+	return b
+}
+
+// Checks if two bit arrays are equal
+func (b *BitArray) Equal(x *BitArray) bool {
+	// TODO(weiihann): this is really not a good thing to do...
+	if b == nil && x == nil {
+		return true
+	} else if b == nil || x == nil {
+		return false
+	}
+
+	return b.len == x.len &&
+		b.words[0] == x.words[0] &&
+		b.words[1] == x.words[1] &&
+		b.words[2] == x.words[2] &&
+		b.words[3] == x.words[3]
+}
+
+// Returns true if bit n-th is set, where n = 0 is LSB.
+// The n must be <= 255.
+func (b *BitArray) IsBitSet(n uint8) bool {
+	if n >= b.len {
+		return false
+	}
+
+	return (b.words[n/64] & (1 << (n % 64))) != 0
+}
+
+// Serialises the BitArray into a bytes buffer in the following format:
+// - First byte: length of the bit array (0-255)
+// - Remaining bytes: the necessary bytes included in big endian order
+// Example:
+//
+//	BitArray{len: 10, words: [4]uint64{0x03FF}} -> [0x0A, 0x03, 0xFF]
+func (b *BitArray) Write(buf *bytes.Buffer) (int, error) {
+	if err := buf.WriteByte(b.len); err != nil {
+		return 0, err
+	}
+
+	n, err := buf.Write(b.activeBytes())
+	return n + 1, err
+}
+
+// Deserialises the BitArray from a bytes buffer in the following format:
+// - First byte: length of the bit array (0-255)
+// - Remaining bytes: the necessary bytes included in big endian order
+// Example:
+//
+//	[0x0A, 0x03, 0xFF] -> BitArray{len: 10, words: [4]uint64{0x03FF}}
+func (b *BitArray) UnmarshalBinary(data []byte) {
+	b.len = data[0]
+
+	var bs [32]byte
+	copy(bs[32-b.byteCount():], data[1:])
+	b.setBytes32(bs[:])
+}
+
+// Sets b to the same value as x.
+func (b *BitArray) Set(x *BitArray) *BitArray {
+	b.len = x.len
+	b.words[0] = x.words[0]
+	b.words[1] = x.words[1]
+	b.words[2] = x.words[2]
+	b.words[3] = x.words[3]
+	return b
+}
+
+// Sets b to the bytes representation of a felt.
+func (b *BitArray) SetFelt(length uint8, f *felt.Felt) *BitArray {
+	b.setFelt(f)
+	b.len = length
+	return b
+}
+
+// Sets b to the bytes representation of a felt with length 251.
+func (b *BitArray) SetFelt251(f *felt.Felt) *BitArray {
+	b.setFelt(f)
+	b.len = 251
+	return b
+}
+
+// Interprets the data as the big-endian bytes, sets b to that value and returns b.
+// If the data is larger than 32 bytes, only the first 32 bytes are used.
+func (b *BitArray) SetBytes(length uint8, data []byte) *BitArray {
+	b.setBytes32(data)
+	b.len = length
+	return b
+}
+
+// Sets b to the uint64 representation of a bit array.
+func (b *BitArray) SetUint64(length uint8, data uint64) *BitArray {
+	b.words[0] = data
+	b.len = length
```

Comment:
Can the length be inferred from data? What happens if length doesn't represent the data?

Reply:
`length` cannot be inferred from the `data`. For example:

Given uint64 of `0b11111111` (8 bits) and length of `10`, this means that the actual bit array is illustrated as `0b0011111111`.

Reply:
Should there be a check when length is < 8 bits? Since it is clearly non intended by the programmer we can panic at than instance. What do you think?

Reply:
`length` < 8 bits, this is a valid input.

Reply:
Given a uint64 `0b11111111` would length < 8 be still valid?

Reply:
Hmm in that case it may cause some unintended behaviours (for example, `rsh`). My latest changes now clear the unused bits and ensure that they are zeros.